File: master/add_binary.cpp
#include "header.h"

string addBinary(string a, string b) {
    string result;
    int i=a.size()-1, j=b.size()-1;
    int carry = 0;
    while(i>=0 || j>=0){
        int a_curr = i>=0?(int)(a[i]-'0'):0;
        int b_curr = j>=0?(int)(b[j]-'0'):0;
        result = (char)((a_curr+b_curr+carry)%2+'0')+result;
        carry = (a_curr+b_curr+carry)>1?1:0;
        if(i>=0) i--;
        if(j>=0) j--;
    }
    if(carry)
        result = (char)(carry+'0')+result;
    return result;
}

int main(){
    string a("11");
    string b("1");
    string res = addBinary(a,b);
    printf("%s\n", res.c_str());
    return 0;
}

File: master/add_binary.py
import itertools

class Solution:
    # @param a, a string
    # @param b, a string
    # @return a string
    def addBinary(self, a, b):
        res = ""
        carry = 0
        for chars in itertools.izip_longest(reversed(a),reversed(b)):
            a_num = 0 if chars[0] is None else 0 if chars[0]=='0' else 1
            b_num = 0 if chars[1] is None else 0 if chars[1]=='0' else 1
            x = a_num+b_num+carry
            res = ('0' if x%2==0 else '1' )+res
            carry = x/2
        if carry:
            res = '1'+res
        return res



if __name__ == "__main__":
    print Solution().addBinary("11","1")
